### ISSUE TITLE
[HOLD for @0xmiroslav payment] [$2000] Request Money modal closed when selecting currency for new request or edit

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#25626]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Break down in numbered steps

1. Click on FAB
2. Request Money
3. On the amount page, click the currency symbol to change the currency
4. Select some other currency

## Expected Result:
Describe what you think should've happened

You are back on the amount page with selected currency

## Actual Result:
Describe what actually happened

The modal has closed

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** v1.3.56-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**
**Issue reported by:** @parasharrajat 
**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01732c4f8d4d117765</li>
        <li>Upwork Job ID: 1694010840997621760</li>
        <li>Last Price Increase: 2023-08-31</li>
<li>Automatic offers: </li>
<ul>
<li>0xmiroslav | Reviewer | 26452949</li>
<li>shubham1206agra | Contributor | 26452951</li>
</ul></li>
    </ul>
</details>

**Comment by mountiny** [comment]
forgot to include this:

https://github.com/Expensify/App/assets/36083550/86577cfe-f025-499e-946c-1a2c3f1a4beb

**Comment by conorpendergrast** [comment]
I agree that this is a deploy blocker. @mountiny @roryabraham any ideas for which PR caused this?

**Comment by cubuspl42** [comment]
@mountiny @conorpendergrast I reproduced this. I can work on this if you're still looking for someone.

**Comment by mountiny** [comment]
@cubuspl42 We are actively looking into this one, if you have a valid proposal identifying the offending PR (this is a deploy blocker so there must have been a recent change in staging which introduced it), I could let you work on it for $500

**Comment by cubuspl42** [comment]
So far it looks to me like it was always broken with the ("Manual" / "Scan" / "Distance") header, which is under beta. So I guess it's not a release blocker?

**Comment by cubuspl42** [comment]
@mountiny Could you test this with this manual patch?

```js
function canUseScanReceipts(betas) {
    return false; // <--
    return _.contains(betas, CONST.BETAS.SCAN_RECEIPTS) || canUseAllBetas(betas);
}
```

**Comment by Pluto0104** [comment]
@cubuspl42, Of course, it can be only reproduced with "Manual" / "Scan" / "Distance" header.

**Comment by cubuspl42** [comment]
@Pluto0104 But are bugs reproducible only with beta features considered deploy blockers?

**Comment by Pluto0104** [comment]
@cubuspl42, Since I'm new to contributing to this project, I may not have a thorough understanding of the beta features yet. However, based on my current knowledge, I don't believe the bug is severe, frequent, or affecting a significant number of users who rely on the beta feature. Therefore, I don't consider it to be a deploy blocker.

**Comment by cubuspl42** [comment]
@mountiny Anyway, let me know if we degrade this to a non-blocker and if the work offer you proposed stands.

**Comment by mountiny** [comment]
> So far it looks to me like it was always broken with the ("Manual" / "Scan" / "Distance") header, which is under beta. So that would mean it's been broken for weeks. So I guess it's not a release blocker?

The beta worked fine for us for weeks, this is indeed recent

**Comment by mountiny** [comment]
In general, you are right that beta features might not be considered as deploy blockers. At the moment tho the Scan and Distance are our highest priority so we definitely want to fix this asap. It is also making testing of new features locally where the betas are enabled harder

**Comment by mountiny** [comment]
I have confirmed this is indeed repro in production too, and hence it must be related to the beta and most likely to the OnyxTabNavigator which is hidden behind beta for most

**Comment by mountiny** [comment]
given its not a deploy blocker, ideal external issue

**Comment by Piotrfj** [comment]
Hey, I'm Piotr from Callstack - expert contributor group - I would like to help resolve this issue

**Comment by Piotrfj** [comment]
I have started investigating the issue, but im OOO today and tomorrow. Will come back to it on monday

**Comment by roryabraham** [comment]
I see we have interest from multiple expert contractors but for the sake of fairness and efficiency I'm going to await a formal proposal here

**Comment by Pluto0104** [comment]
@cubuspl42, it's possible that this error is caused by nesting the navigation. I could resolve the error, but I haven't identified the underlying issue yet. It appears to be related to managing navigation outside of the Tab Navigation. How do you think?

**Comment by cubuspl42** [comment]
@Pluto0104 I'm not actively involved in this issue anymore, and I'm not the assigned C+. Also if you'd like to start a longer discussion, it will probably be more welcome on Slack.

**Comment by conorpendergrast** [comment]
Looks like we're still waiting for proposals here. @Piotrfj any progress so far?

**Comment by Piotrfj** [comment]
@conorpendergrast found a solution for the problem, will provide proposal soon

**Comment by Piotrfj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem we are trying to solve is the request money modal self closing after changing the currency.

### What is the root cause of that problem?
Navigation not working properly.

### What changes do you think we should make in order to solve the problem?

Change the navigate method from
`Navigation.navigate(`${props.route.params.backTo}?currency=${option.currencyCode}`);`
to
`Navigation.goBack(`${props.route.params.backTo}?currency=${option.currencyCode}`, true);`
inside `IOUCurrencySelection.js` file (line 91)

### What alternative solutions did you explore?
N/A

**Comment by Pluto0104** [comment]
Of course, Navigation is not working properly. Please see [here](https://github.com/Expensify/App/issues/25626#issuecomment-1692245869). But what is the root cause? Can you explain this?

**Comment by s77rt** [comment]
This is a regression from https://github.com/Expensify/App/pull/24165 but that PR is not the root cause, it's a PR that just made the bug reproducible. This seems related to nested navigators. I think the root cause has something to do with navigating to the same page twice.

<img width="1158" alt="Screenshot 2023-08-31 at 8 39 44 PM" src="https://github.com/Expensify/App/assets/16493223/1892e68c-9f69-4d69-9b7c-e9018b7fa4d7">

**Comment by mountiny** [comment]
Seems like we got a proposal coming from Pluto which should work https://expensify.slack.com/archives/C01GTK53T8Q/p1693514634556789?thread_ts=1693510185.535349&cid=C01GTK53T8Q, once confirmed, lets make a PR

**Comment by s-alves10** [comment]
I think we need to find the RCA and fix it. @Piotrfj 's solution looks like a workaround

**Comment by mountiny** [comment]
Agreed, that solution does not work fully either

**Comment by napster125** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Request Money modal closed when selecting currency.
### What is the root cause of that problem?
The root cause is the default behavior of the Tab Navigator.
- When we navigate to a Tab Navigator, the default behavior is to render the initial component of the selected tab and also set the focus on the tab bar.
- As a result of the focus being on the tab bar, any modals or other UI elements that previously had focus might lose focus. This can lead to unexpected behavior, such as modals being dismissed due to the focus shift caused by the tab bar.
- The [comment](https://github.com/Expensify/App/issues/25626#issuecomment-1701684107) indicates that when we navigate to the tab navigator, the first navigation involves manually navigating to a specific URL using the navigationContainerRef. This likely occurs before the Tab Navigator's default behavior takes over.

### What changes do you think we should make in order to solve the problem?
This problem is only occurred when we navigate to the tab navigator, so I suggest modifying the code at the specified line to handle the navigation differently based on the scenario.
https://github.com/Expensify/App/blob/2812e45d725d7ee4ef988a2217853a718fb7f3b9/src/pages/iou/IOUCurrencySelection.js#L87


If the previous navigation state is associated with a 'tab' type, we're using the tab navigation approach, utilizing the props navigation and navigating to the appropriate tab within the tab navigator.
If the previous navigation state does not correspond to the 'tab' type (i.e., navigation outside the tab navigator), we're still using the navigationContainerRef approach to navigate.

For example:
```
prevState && prevState.type === 'tab'
                    ? props.navigation.navigate(prevState.routeNames[prevState.index], {iouType, reportID, currency: option.currencyCode})
                    : Navigation.navigate(`${props.route.params.backTo}?currency=${option.currencyCode}`);
```

This implementation effectively manages the navigation behavior based on the context, providing the correct navigation approach depending on whether we are navigating within the tab navigator or not. It also maintains compatibility with other cases where the navigationContainerRef approach works well.

### What alternative solutions did you explore? (Optional)
I have investigated potential alternative solutions, one of which involved using the following approach:
```
Navigation.navigate(`${props.route.params.backTo}?currency=${option.currencyCode}`, CONST.NAVIGATION.TYPE.UP); 
```
However, this alternative solution resulted in an undesired outcome. Specifically, when a new currency is selected, it triggered a reset of the amount. Subsequently, when you edited the amount in the confirmation list, the absence of a back button prevented you from reverting the changes. This situation led to an undesirable user experience.

**Comment by shubham1206agra** [comment]
Can anyone confirm the reproducibility of this issue after applying the patch from https://github.com/Expensify/App/pull/25989?
cc @mountiny @parasharrajat

**Comment by s-alves10** [comment]
This isn't reproducible now.

**Comment by shubham1206agra** [comment]
@s-alves10 Before patch or after patch?

**Comment by s-alves10** [comment]
After patch

**Comment by mountiny** [comment]
I have tried that before actually as i had the same feeling as you @shubham1206agra but i could reproduce still :/

**Comment by getusha** [comment]
are we still looking for another proposals?

**Comment by mountiny** [comment]
@napster125 thanks! How does this fix the edit modal or the confirmation page currency selector? There is not tab navigator in those pages?

@0xmiroslav what do you think about the proposal?

**Comment by shubham1206agra** [comment]
@mountiny I was asking cause I wasn't able to repro after the patch

**Comment by shubham1206agra** [comment]
Can you help me to repro this?

**Comment by shubham1206agra** [comment]
> > Can you help me to repro this?
> 
> @shubham1206agra This is 100% reproducible. Did you enable `canUseScanReceipts` or `canUseDistanceRequests` beta permission?

Yes
Able to repro on main

https://github.com/Expensify/App/assets/58412969/b77032df-5bf4-41bd-88f8-1c7727ba9afb

**Comment by shubham1206agra** [comment]
But not after applying the patch

https://github.com/Expensify/App/assets/58412969/9e915917-4da5-401d-9fef-5bb3832163d2

**Comment by shubham1206agra** [comment]
Is there anything I am doing wrong here?

**Comment by shubham1206agra** [comment]
See https://github.com/Expensify/App/pull/25989

**Comment by tamdao** [comment]
@shubham1206agra I have applied the patch https://github.com/Expensify/App/pull/25989 and can't preproduce this issue

**Comment by mountiny** [comment]
Hired @shubham1206agra here, the patch works and its going to staging as its fixing couple of issues and problem in our core flow.

**Comment by shubham1206agra** [comment]
Hey @mountiny 
Looks like there may be some focusing issues.
Should I take a look at them in this issue only?

**Comment by mountiny** [comment]
@shubham1206agra can you elaborate on the issues? Are they happening after applying the patch only?

**Comment by shubham1206agra** [comment]
Nope before the patch too
Only on safari web in mac

**Comment by mountiny** [comment]
@shubham1206agra if its a standalone issue and its was before the patch, then we should log it as a separate issue, could you log it with reproduction steps? It might be that is browser specific issue which are lower priority, Chrome is the main focus

**Comment by shubham1206agra** [comment]
Ok

**Comment by conorpendergrast** [comment]
@abekkala I'm on parental leave; re-assigning! This is set for payment today and a review to see if @parasharrajat should be [considered as reporter too](https://github.com/Expensify/App/issues/25626#issuecomment-1708950663)

**Comment by abekkala** [comment]
PAYMENT DETAILS: 

Reporter: Vit 
@shubham1206agra  [offer](https://www.upwork.com/nx/wm/offer/26452951) (Contributor who fixed)
@0xmiroslav [offer](https://www.upwork.com/nx/wm/offer/26452949) (PR Reviewer)

@mountiny I also reviewed the PR - can you confirm the delay was due to QA an bonus is warranted?

**Comment by mountiny** [comment]
@parasharrajat is a reported I believe

This was linked to multiple issues and we did not get this merged that fast, not saying by contirbutors mistake, but I believe the urgency bonus should be really applied when its merged fast as a collective effort.

I do not see @0xmiroslav as a reviewer here, @0xmiroslav can you confirm your work on this issue please? thanks!

**Comment by mountiny** [comment]
Thanks

**Comment by abekkala** [comment]
Payments for today: 

Reporter: @parasharrajat [$50]
@shubham1206agra [offer](https://www.upwork.com/nx/wm/offer/26452951) (Contributor who fixed) [$2000 - no bonus]
@0xmiroslav [offer](https://www.upwork.com/nx/wm/offer/26452949) (PR Reviewer) [$2000 - no bonus]

@0xmiroslav can you complete the checklist above so I can complete payments?

**Comment by abekkala** [comment]
@shubham1206agra - paid and contract ended! 🎉 Thank You!

@0xmiroslav can you **accept** the offer, in Upwork, so I can submit payment: https://www.upwork.com/nx/wm/offer/26452949

@parasharrajat I've sent you a Reporting Payout offer in upwork: https://www.upwork.com/nx/wm/offer/26616784

**Comment by abekkala** [comment]
@parasharrajat ah yes, I see now that you get paid via Expensify Chat

**Comment by abekkala** [comment]
not overdue - waiting on further guidance on payment

**Comment by JmillsExpensify** [comment]
$250 payment for @parasharrajat approved based on BZ summary.

**Comment by abekkala** [comment]
STATUS:
Waiting on @0xmiroslav.
https://expensify.slack.com/archives/C01SKUP7QR0/p1695166357917099?thread_ts=1694206723.155449&cid=C01SKUP7QR0

Please ping me here when you're able to accept payments in Upwork.

**Comment by abekkala** [comment]
@0xmiroslav are you ready to accept payments via Upwork yet?
